text/3857-cfg-version.md

```diff
@@ -0,0 +1,609 @@
+
+Otherwise, the `IDENTIFIER` will be compared to the string literal according to
+[Cargo's `>=` version requirements](https://doc.rust-lang.org/nightly/cargo/reference/specifying-dependencies.html#comparison-requirements).
+For example, `#[cfg(since(rust, "1.90"))]` will be treated as `1.95.2 >= 1.90.0`.
+
+See also `--check-cfg`.
+
+## `--check-cfg`
+
+A new predicate will be added of the form:
+```
+ConfigurationVersion -> `since` `(` ( STRING_LITERAL | RAW_STRING_LITERAL ) `)`
+```
+
+The syntax for the contents of the string literal is a SemVer value without the build field.
+
+This will specify that the given `--cfg` is valid for all values that match:
+- SemVer syntax
+- from the specified version and up
+
+When the given `--cfg` is used with the `since` predicate:
+- the string literal should not be of a syntax that evaluates to `false`
+- the minimum version requirement must specify a subset of what the `--check-cfg` specifies
+
+So given `--check-cfg 'cfg(foo, values(since("1.95.0")))'`,
+- ✅ `#[cfg(foo = "1.100.0")]`
+- ⚠️ `#[cfg(foo = "1.0")]`: not SemVer syntax
+- ✅ `#[cfg(since(foo, "1.95.0"))]`
+- ✅ `#[cfg(since(foo, "1.100.0"))]`
+- ✅ `#[cfg(since(foo, "3.0.0"))]`
+- ✅ `#[cfg(since(foo, "1.95"))]`
+- ⚠️ `#[cfg(since(foo, "1.90.0"))]`: matches a superset of `--check-cfg`
+- ⚠️ `#[cfg(since(foo, "1.95.0-0"))]`: matches a superset of `--check-cfg`
+- ⚠️ `#[cfg(since(foo, "1"))]`: matches a superset of `--check-cfg`
+- ⚠️ `#[cfg(since(foo, "bar"))]`: invalid string literal syntax
+
+## `rust` cfg
+
+A new built-in cfg `--cfg rust=<version>` will be added by the compiler
+that specifies the language version.
+This will be the version of `rustc` with the behavior for pre-release versions being unspecified.
+We expect rustc to strip pre-release versions, treating them as if they fully implement that language version.
+
+`rust` will be specified as `--check-cfg 'cfg(rust, values(since("1.95.0")))'`
+(or whatever version this gets stabilized in).
+
+This will be reported back through `--print-cfg`.
+
+Because this gets reported back in `--print-cfg`,
+Cargo will expose `rust` in:
+- build scripts as `CARGO_CFG_RUST`
+- `target."cfg()".dependencies`
+
+## clippy
+
+Clippy has a [`clippy::incompatible_msrv`](https://rust-lang.github.io/rust-clippy/master/index.html#incompatible_msrv) lint
+which will fire whenever a standard library item is used with a `#[stable(since)]` newer than `package.rust-version`.
+However, it will be perfectly reasonable to use those items when guarded by a `#[cfg(since)]`.
+
+Clippy may wish to:
+- Find a way to reduce false positives, e.g. evaluating the `cfg(since)`s that led to the item's usage or disabling the lint within `#[cfg(since)]`
+- Suggest `#[cfg(since)]` in the `clippy::incompatible_msrv` diagnostic report (maybe along with offering to bump MSRV)
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+People may be using `--cfg rust` already and would be broken by this change.
```

Comment:
At least with my [initial search](https://grep.app/search?f.lang=Rust&f.lang.pattern=rust&case=true&regexp=true&q=cfg.*%5Cbrust%5Cb), I did not see any public uses